src/ui/FocusTrap/FocusTrap.ts

```diff
@@ -0,0 +1,112 @@
+import _ = require('underscore');
+import { Defer } from '../../misc/Defer';
+
+export class FocusTrap {
+  private focusInEvent: (e: FocusEvent) => void;
+  private focusOutEvent: (e: FocusEvent) => void;
+  private enabled: boolean;
+
+  private get focusableElements(): HTMLElement[] {
+    return _.sortBy(this.container.querySelectorAll('[tabindex]'), element => element.tabIndex);
+  }
+
+  constructor(private container: HTMLElement) {
+    this.enable();
+  }
+
+  public disable() {
+    document.removeEventListener('focusin', this.focusInEvent);
+    document.removeEventListener('focusout', this.focusOutEvent);
+    this.enabled = false;
+  }
+
+  private enable() {
+    document.addEventListener('focusin', (this.focusInEvent = e => this.onFocusIn(e)));
+    document.addEventListener('focusout', (this.focusOutEvent = e => this.onFocusOut(e)));
+    this.enabled = true;
+  }
+
+  private getSibling(element: HTMLElement, previous = false) {
+    const elements = this.focusableElements;
+    const currentIndex = elements.indexOf(element);
+    if (currentIndex === -1) {
+      return null;
+    }
+    return elements[(currentIndex + (previous ? -1 : 1) + elements.length) % elements.length];
+  }
+
+  private focusSibling(element: HTMLElement, previous = false) {
+    const sibling = this.getSibling(element, previous);
+    if (sibling) {
+      sibling.focus();
+    }
+  }
+
+  private focusFirstElement() {
+    const elements = this.focusableElements;
+    if (elements.length > 0) {
```

Comment:
could only use `if (elements.length)`

Reply:
Yeah that's what I initially wrote but I wasn't sure which option was preferable. I'll change it to `elements.length`.